### PR TITLE
Large message v4

### PR DESCRIPTION
**Pull request by juimonen**

### CODE REVIEW

**Comment by juimonen** [comment]
related driver change https://github.com/thesofproject/linux/pull/723

**Comment by juimonen** [comment]
@lgirdwood elems_remaining is now there

File: src/audio/eq_fir.c [hunk 1]
```diff
@@ -484,21 +484,37 @@ static int fir_cmd_get_data(struct comp_dev *dev,
...
-			if (bs > SOF_EQ_FIR_MAX_SIZE || bs == 0 ||
-			    bs > max_size)
-				return -EINVAL;
-			memcpy(cdata->data->data, cd->config, bs);
+			cdata->elems_remaining = 0;
+			offset = 0;
+			if (bs > max_size) {
+				bs = (cdata->msg_index + 1) * max_size > bs ?
+					bs - cdata->msg_index * max_size :
+				  max_size;
```
Comment (abonislawski) [inline src/audio/eq_fir.c]: at least fix the indent

File: src/audio/eq_fir.c [hunk 2]
```diff
@@ -484,21 +484,37 @@ static int fir_cmd_get_data(struct comp_dev *dev,
...
+				offset = cdata->msg_index * max_size;
+				cdata->elems_remaining = cd->config->size -
+					offset;
+			}
+			cdata->num_elems = bs;
+			trace_eq("fir_cmd_get_data(), %zu %u %u %zu",
```
Comment (abonislawski) [inline src/audio/eq_fir.c]: print more info about variables, like on line 589 ("blob size: %u" etc.)

File: src/audio/eq_fir.c [hunk 3]
```diff
@@ -484,21 +484,37 @@ static int fir_cmd_get_data(struct comp_dev *dev,
...
+				 bs, cdata->msg_index, max_size, offset);
+			memcpy(dst, src + offset, bs);
```
Comment (abonislawski) [inline src/audio/eq_fir.c]: We should start using memcpy_s now or wait for pr #1103 ?

File: src/audio/eq_fir.c (hunk 1 above)
Comment (juimonen) [reply src/audio/eq_fir.c]: ok

File: src/audio/eq_fir.c (hunk 2 above)
Comment (juimonen) [reply src/audio/eq_fir.c]: ok

File: src/audio/eq_fir.c (hunk 3 above)
Comment (juimonen) [reply src/audio/eq_fir.c]: whatever suits better...

File: src/audio/eq_fir.c [hunk 4]
```diff
@@ -569,28 +586,43 @@ static int fir_cmd_set_data(struct comp_dev *dev,
...
+		    SOF_EQ_FIR_MAX_SIZE)
 			return -EINVAL;
 
-		/* Allocate buffer for copy of the blob. */
-		cd->config = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, bs);
-		if (!cd->config) {
-			trace_eq_error("fir_cmd_set_data() error: "
-				       "buffer allocation failed");
-			return -EINVAL;
+		if (cdata->msg_index == 0) {
```
Comment (lgirdwood) [inline src/audio/eq_fir.c]: Should we also check elems_remaining here as msg->index will be 0 by default on small messages too ?

**Comment by plbossart** [comment]
> I'm happy to merge this when ready as long as IPC is agreed with kernel. Kernel part requires some extra work with IPC operations.

I am actually having second thoughts on all this. It's a bit odd to enable support for large messages in the IPC, but make it up to the platform driver to decide to use it or not but without the ability to discover what the firmware actually supports and what the mailbox size actually is.

**Comment by lgirdwood** [comment]
> > I'm happy to merge this when ready as long as IPC is agreed with kernel. Kernel part requires some extra work with IPC operations.
> 
> I am actually having second thoughts on all this. It's a bit odd to enable support for large messages in the IPC, but make it up to the Linux kernel platform-specific driver to decide to use it or not but without the ability to discover what the firmware actually supports and what the mailbox size actually is.

The Kernel platform drivers know the mailbox limits via the extended FW ready data (sent on FW boot). This contains runtime information from the FW that can be used by kernel platform driver init code.

**Comment by juimonen** [comment]
@lgirdwood should this be merged now?

**Pull request merged**